Sming/Components/OtaUpgrade/OtaUpgradeStream.h

```diff
@@ -0,0 +1,209 @@
+#ifdef OTA_SIGNED
+#include <sodium/crypto_sign.h>
+#endif
+#ifdef OTA_ENCRYPTED
+#include <sodium/crypto_secretstream_xchacha20poly1305.h>
+#endif
+
+/**
+ * @brief Parse and apply firmware upgrade files generated by otatool.py
+ * 
+ * A Write-only stream that parses and processed firmware upgrade files generated
+ * by otatool.py.
+ *
+ * The class fully automates the firmware upgrade process without any manual 
+ * configuration. At construction time, the rBoot configuration is read to determine 
+ * the unused ROM slot which should receive the upgrade.
+ * Just feed the upgrade file content into the `write()` method in arbitrarily 
+ * sized chunks. The relevant portion(s) of the Flash memory (currently only the 
+ * application rom) are updated on the fly as data arrives. When the file is complete
+ * and signature validation (if enabled) was successful, the updated slot is activated
+ * in the rBoot configuration.
+ * Call `hasError()` and/or check the public \c #errorCode member to determine if 
+ * everything went smoothly.
+ * 
+ * For further information on configuration options and the file format, 
+ * refer to the component's documentation.
+
+ * @note: If encryption is enabled, the class dynamically allocates a buffer to fit
+ * the largest chunk of the encryption container (2kB unless otatool.py was modified).
+ * Otherwise, no dynamic memory allocation is performed.
+ */
+class OtaUpgradeStream : public ReadWriteStream
+{
+public:
+	OtaUpgradeStream();
+	~OtaUpgradeStream();
+
+	enum ErrorCode {
+		NoError, ///< No error occured thus far (default value of \c #errorCode if `hasError()` returns false)
+		InvalidFormatError, ///< Invalid/unsupported upgrade file format
+		UnsupportedDataError, ///< Some content of the upgrade file is not supported by this version of OtaUpgradeStream.
+		DecryptionError, ///< Decryption failed. Probably wrong decryption key.
+		NoRomFoundError, ///< The file did not contain a ROM image suitable for the start address of the slot to upgrade.
+		RomTooLargeError, ///< The contained ROM image does not fit into the application firmware slot.
+		DowngradeError, ///< Attempt to downgrade to older firmware version.
+		VerificationError, ///< Signature verification failed - updated ROM not activated
+		FlashWriteError, ///< Error while writing to Flash memory.
+		RomActivationError, ///< Error while activating updated ROM slot.
+		OutOfMemoryError, ///< Dynamic memory allocation failed
+		InternalError ///< An unexpected error occured.
+	};
+
+	ErrorCode errorCode = NoError; ///< Error code. Only relevant if `hasError()` returns true.
+
+	/** @brief Convert error code to string.
+	 * @see `errorCode`
+	 */
+	static String errorToString(ErrorCode code);
+
+	/** @brief Process chunk of upgrade file
+	 * @param data Pointer to chunk of data.
+	 * @param size Size of chunk pointed to by `data` in bytes.
+	 * @return If less than `size`, an error occured. Check \c #errorCode for more details.
```

Comment:
As far as I know, `@retval` is intended to describe a particular return value and is thus only suited for bools, enums, etc., while `@return` is used for more general descriptions, e.g.
```
@retval true the function succeeded
@retval false the function failed
```
as opposed to
```
@return \c true if the function succeeded and \c false otherwise.
```
In the particular case above, I  think `@return` is the right choice, although I have to admit I'm just used to it and did not think too much about it.